### ISSUE TITLE
Not clear exception

### ISSUE THREAD

**Issue by ZhukovGreen** [MSLNZ/msl-loadlib#6]
HI, I am trying to connect not net assembly and having the following result:
```
_____________________________ test_msl_connector ______________________________

    def test_msl_connector():
        """Get Axapta object via msl-loadlib package."""
>       connectorLL = LoadLibrary(DLL_PATH, 'net')

test_main.py:12:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <[AttributeError("'LoadLibrary' object has no attribute '_lib'") raised in repr()] LoadLibrary object at 0
x4152c18>
path = WindowsPath('C:/Users/AZ/Desktop/test_bom-mcs/Microsoft.Dynamics.BusinessConnectorNet.dll')
libtype = 'net'

    def __init__(self, path, libtype='cdll'):

        _path = path

        # a reference to the .NET Runtime Assembly
        self._assembly = None

        # assume a default extension if no extension was provided
        if not os.path.splitext(_path)[1]:
            _path += DEFAULT_EXTENSION

        self._path = os.path.abspath(_path)
        if not os.path.isfile(self._path):
            # for find_library use the original 'path' value since it may be a library name
            # without any prefix like lib, suffix like .so, .dylib or version number
            self._path = ctypes.util.find_library(path)
            if self._path is None:  # then search sys.path and os.environ['PATH']
                success = False
                search_dirs = sys.path + os.environ['PATH'].split(os.pathsep)
                for directory in search_dirs:
                    p = os.path.join(directory, _path)
                    if os.path.isfile(p):
                        self._path = p
                        success = True
                        break
                if not success:
                    raise IOError('Cannot find the shared library "{}"'.format(path))

        if libtype == 'cdll':
            self._lib = ctypes.CDLL(self._path)
        elif libtype == 'windll':
            self._lib = ctypes.WinDLL(self._path)
        elif libtype == 'oledll':
            self._lib = ctypes.OleDLL(self._path)
        elif libtype == 'net' and self.is_pythonnet_installed():
            import clr
            try:
                # By default, pythonnet can only load libraries that are for .NET 4.0+.
                #
                # When MSL-LoadLib is installed, the useLegacyV2RuntimeActivationPolicy
                # property should have been enabled automatically to allow for loading
                # assemblies from previous .NET Framework versions.
                self._assembly = clr.System.Reflection.Assembly.LoadFile(self._path)

            except clr.System.IO.FileLoadException as err:
                # Example error message that can occur if the library is for .NET <4.0,
                # and the useLegacyV2RuntimeActivationPolicy is not enabled:
                #
                # " Mixed mode assembly is built against version 'v2.0.50727' of the
                #  runtime and cannot be loaded in the 4.0 runtime without additional
                #  configuration information. "
                #
                # To solve this problem, a <python-executable>.config file must exist and it must
                # contain a useLegacyV2RuntimeActivationPolicy property that is set to be "true".
                if "Mixed mode assembly" in str(err):
                    status, msg = self.check_dot_net_config(sys.executable)
                    if not status == 0:
                        raise IOError(msg)
                    else:
                        update_msg = 'Checking .NET config returned "{}" '.format(msg)
                        update_msg += 'and still cannot load library.\n'
                        update_msg += str(err)
                        raise IOError(update_msg)
                raise IOError('The above "System.IO.FileLoadException" is not handled.\n')

            # the shared library must be available in sys.path
            head, tail = os.path.split(self._path)
            sys.path.insert(0, head)

            # don't include the library extension
            clr.AddReference(os.path.splitext(tail)[0])

            # import namespaces, create instances of classes or reference a System.Type[] object
            dotnet = {}
>           for t in self._assembly.GetTypes():
E           System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Ret
rieve the LoaderExceptions property for more information.
E              at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
E              at System.Reflection.Assembly.GetTypes()

..\..\.virtualenvs\test_bom-mcs-bwfslhpz\lib\site-packages\msl\loadlib\load_library.py:132: ReflectionTypeLoadExc
eption
```
I can't understand it. What happens?

**Comment by jborbely** [comment]
Thanks for reporting this issue.

Indeed. The error message is unclear because it results from nested exceptions.

The final exception `AttributeError("'LoadLibrary' object has no attribute '_lib'") raised in repr()` can be easily fixed by defining `self._lib=None` before the library is attempted to be loaded. I will make this change to the code.

The initial cause of the error comes from `System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information`. A reason for this error occurring is due to the fact that I currently do not have a large collection of .NET libraries to test the code and so I have not come across this error message yet.

One solution that I can implement is to wrap the section of the code that walks through the data types in the .NET library in a generic **try..except** block to ignore any loading-time error and allow the user to access those .NET types via the `LoadLibrary.assembly` property rather than including those data types in the `msl.loadlib.load_library.DotNetContainer` object that is returned by `LoadLibrary.lib`.

My preferred solution would be to use a specific **try..except** block as suggested [here](https://stackoverflow.com/a/6086719). However, I do not have access to the `Microsoft.Dynamics.BusinessConnectorNet.dll` file to test this suggestion. Is it possible for you to share this dll (and it's own dependencies) with me so that I can reproduce and fix the error?

**Comment by ZhukovGreen** [comment]
Thx for clarification, @jborbely
I made a simple repo to show the problem https://gitlab.com/remak_team/open-source/MicrosoftDynamicsConnector_python_wrapper

**Comment by jborbely** [comment]
I cannot reproduce the error on my computer; however, I am able to instantiate the Axapta class.

```python
Python 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 10:22:32) [MSC v.1900 64 bit (AMD64)] on win32
In[2]: from msl import loadlib
In[3]: loadlib.version_info
Out[3]: version_info(major=0, minor=3, micro=2)
In[4]: net = loadlib.LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll', 'net')
```
The following returns the **Microsoft.Dynamics.BusinessConnectorNet** Namespace as a Python module
```python
In[5]: Microsoft = getattr(net.lib, 'Microsoft.Dynamics.BusinessConnectorNet')
In[6]: Microsoft
Out[6]: <module 'Microsoft'>
```
So we can now access the Axapta class using
```python
In[7]: Axapta = Microsoft.Dynamics.BusinessConnectorNet.Axapta
In[8]: Axapta
Out[8]: Microsoft.Dynamics.BusinessConnectorNet.Axapta
```
After modifying `test_main.py` in your repo to reflect these changes I was able to run pytest with success. Here is the modified `test_main.py`
```python
"""Implements wrapper for axapta bussiness connector."""

import pathlib
from msl.loadlib import LoadLibrary
import clr

DLL_PATH = pathlib.Path(__file__).parent / 'Microsoft.Dynamics.BusinessConnectorNet.dll'


def test_msl_connector():
    """Get Axapta object via msl-loadlib package."""
    connectorLL = LoadLibrary(DLL_PATH, 'net')

    # this returns the Microsoft.Dynamics.BusinessConnectorNet Namespace
    # as a Python module
    Microsoft = getattr(connectorLL.lib,
                     'Microsoft.Dynamics.BusinessConnectorNet')
    # so we can now access the Axapta class using
    return Microsoft.Dynamics.BusinessConnectorNet.Axapta


def test_pure_pythonnet_connector():
    """Get Axapta object via pythonnet package."""
    clr.AddReference(str(DLL_PATH))
    from Microsoft.Dynamics.BusinessConnectorNet import Axapta
    return Axapta
```
One thing that I did notice is that the Assembly is for .Net Framework v3.5. This means that there must be a file called `python.exe.config` that is located in the same directory as the `python.exe` executable that you are using to run the script. MSL-LoadLib should automatically create this config file for you and then prompt you to run the script again. Check that there is a `python.exe.config` file with
```xml
<configuration>
    <startup useLegacyV2RuntimeActivationPolicy="true">
        <supportedRuntime version="v4.0" />
        <supportedRuntime version="v2.0.50727" />
    </startup>
</configuration>
```
and let me know if you are able to instantiate the Axapta class.

**Comment by ZhukovGreen** [comment]
Hi @jborbely, thank you for your help.

python.exe.config was automatically added.

It seems the issue in the interpreter. If I run the script inside the venv, it gives me an error:
Script:
```python
"""Implements wrapper for axapta bussiness connector."""

import pathlib
from msl.loadlib import LoadLibrary

DLL_PATH = pathlib.Path(
    __file__).parent / 'Microsoft.Dynamics.BusinessConnectorNet.dll'


def test_msl_connector():
    """Get Axapta object via msl-loadlib package."""
    connectorLL = LoadLibrary(DLL_PATH, 'net')
    Microsoft = getattr(connectorLL.lib,
                        'Microsoft.Dynamics.BusinessConnectorNet')
    Axapta = Microsoft.Dynamics.BusinessConnectorNet.Axapta
    return Axapta
```
Result:
```python
>           for t in self._assembly.GetTypes():
E           System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Ret
rieve the LoaderExceptions property for more information.
E              at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
E              at System.Reflection.Assembly.GetTypes()

..\..\.virtualenvs\test_bom-mcs-bwfslhpz\lib\site-packages\msl\loadlib\load_library.py:132: ReflectionTypeLoadException
```

Now test it in anaconda env:
```python
Python 3.6.4 |Anaconda custom (64-bit)| (default, Jan 16 2018, 10:22:32) [MSC v.
1900 64 bit (AMD64)]
Type 'copyright', 'credits' or 'license' for more information
IPython 6.1.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from msl import loadlib

In [2]: loadlib.version_info
Out[2]: version_info(major=0, minor=3, micro=2)

In [3]: net = loadlib.LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll'
   ...: , 'net')
---------------------------------------------------------------------------
ReflectionTypeLoadException               Traceback (most recent call last)
<ipython-input-3-b592e1ecbd9f> in <module>()
----> 1 net = loadlib.LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll',
 'net')

C:\ProgramData\Anaconda3\lib\site-packages\msl\loadlib\load_library.py in __init
__(self, path, libtype)
    130             # import namespaces, create instances of classes or reference a System.Type[] object
    131             dotnet = {}
--> 132             for t in self._assembly.GetTypes():
    133                 if t.Namespace is not None:
    134                     if t.Namespace not in dotnet:

ReflectionTypeLoadException: Unable to load one or more of the requested types.
Retrieve the LoaderExceptions property for more information.
   at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
   at System.Reflection.Assembly.GetTypes()
```

It looks like the same thing.

Any ideas, what may be wrong?

**Comment by ZhukovGreen** [comment]
If it is not complicated, could to please share the result of a dependency walker on this Dll. Maybe this could give us some insights.

Thank you

**Comment by jborbely** [comment]
Here's the output from Dependency Walker:
[Microsoft.Dynamics.BusinessConnectorNet.txt](https://github.com/MSLNZ/msl-loadlib/files/1708564/Microsoft.Dynamics.BusinessConnectorNet.txt)

As another suggestion, remove MSL-LoadLib from the current issue.

Copy this script to the same directory as **Microsoft.Dynamics.BusinessConnectorNet.dll**, run it and let me know if you get the same error. I do not get any errors running this script on a computer with the Dependency Walker results above and with **pythonnet v2.3.0** installed.
```python
"""
Only use clr to load Microsoft.Dynamics.BusinessConnectorNet.dll
"""
import os
import clr

filename = 'Microsoft.Dynamics.BusinessConnectorNet.dll'
path = os.path.abspath(os.path.join(os.path.dirname(__file__), filename))

assembly = clr.System.Reflection.Assembly.LoadFile(path)

clr.AddReference(os.path.splitext(filename)[0])

for t in assembly.GetTypes():
    print('Namespace: {} || Name: {}'.format(t.Namespace, t.Name))
```
Basically, this is what MSL-LoadLib is doing behind the scenes (and, as a side note, it is the `assembly.GetTypes()` loop that caused issue #7).

**Comment by jborbely** [comment]
It seems like the pythonnet developers are currently discussing a similar [issue](https://github.com/pythonnet/pythonnet/pull/528#pullrequestreview-94949974)

**Comment by ZhukovGreen** [comment]
Run it on available Windows Server 2016 machine and it works like a charm:)
On my machine:
```python
platform win32 -- Python 3.6.4, pytest-3.4.0, py-1.5.2, pluggy-0.6.0 -- c:\users\az\.virtualenvs\test_bom-mcs-bwf
slhpz\scripts\python.exe
cachedir: .pytest_cache
rootdir: C:\Users\AZ\Desktop\test_bom-mcs, inifile:
collected 2 items                                                              

test_main.py::test_using_clr_only FAILED

================================== FAILURES ===================================
_____________________________ test_using_clr_only _____________________________

    def test_using_clr_only():
        """
        Only use clr to load Microsoft.Dynamics.BusinessConnectorNet.dll
        """
        import os
        import clr
    
        filename = 'Microsoft.Dynamics.BusinessConnectorNet.dll'
        path = os.path.abspath(os.path.join(os.path.dirname(__file__), filename))
    
        assembly = clr.System.Reflection.Assembly.LoadFile(path)
    
        clr.AddReference(os.path.splitext(filename)[0])
    
>       for t in assembly.GetTypes():
E       System.Reflection.ReflectionTypeLoadException: Unable to load one or more of the requested types. Retriev
e the LoaderExceptions property for more information.
E          at System.Reflection.RuntimeModule.GetTypes(RuntimeModule module)
E          at System.Reflection.Assembly.GetTypes()

test_main.py:32: ReflectionTypeLoadException
============================= 1 tests deselected ==============================
=================== 1 failed, 1 deselected in 0.53 seconds ====================
```

**Comment by ZhukovGreen** [comment]
The output of a dependency walker seems the same as mine

**Comment by jborbely** [comment]
Just for reference, I also ran the test using pipenv instead of a conda env and the test was successful.

Since this issue is most likely a dependency issue in the `Assembly.GetTypes()` routine, I have pushed a change that allows for bypassing the call.

You can test the changes by installing the develop branch
`pip install git+https://github.com/mslnz/msl-loadlib.git@develop`

Hopefully, this means that you can access the `Axapta` class by disabling the problematic `Assembly.GetTypes()` call by passing in `get_assembly_types=False`
```python
>>> from msl.loadlib import LoadLibrary
>>> net = LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll', 'net', get_assembly_types=False)
```
This means that the `net.lib` property value will be `None`
```python
>>> repr(net.lib)
'None'
```
but you can still access the RuntimeAssembly object
```python
>>> repr(net.assembly)
'<System.Reflection.RuntimeAssembly object at 0x00000257642A4978>'
```
and import modules from the library *(note, I also fixed issue #7 in the latest commit)*
```python
>>> from Microsoft.Dynamics.BusinessConnectorNet import Axapta
>>> Axapta
<class 'Microsoft.Dynamics.BusinessConnectorNet.Axapta'>
```

**Comment by ZhukovGreen** [comment]
I think the issue is in dependencies problem, which silently passed.
```python
ipdb> net = LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll', 'net', get_assembly_types=False)
ipdb> repr(net.assembly)
'<System.Reflection.RuntimeAssembly object at 0x000000000689B9E8>'
ipdb> from Microsoft.Dynamics.BusinessConnectorNet import Axapta
*** ModuleNotFoundError: No module named 'Microsoft'
```

If I use clr:
```python
ipdb> import clr
ipdb> clr.AddReference('Microsoft.Dynamics.BusinessConnectorNet.dll')
*** System.IO.FileNotFoundException: Unable to find assembly 'Microsoft.Dynamics.BusinessConnectorNet.dll'.
   at Python.Runtime.CLRModule.AddReference(String name)
```
Perhaps the issue is somewhere in pythonnet, which do not raise an exception

**Comment by jborbely** [comment]
I hope that the following will solve your dependency problem.

I was able to create the Axapta class in Linux, this way none of the Microsoft dependencies can exist so I was able to ensure that all dependencies are accounted for.
```python
Python 3.6.3 |Anaconda, Inc.| (default, Oct 13 2017, 12:02:49) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from msl.loadlib import LoadLibrary
>>> net = LoadLibrary('/home/joe/issue6/Microsoft.Dynamics.BusinessConnectorNet.dll', 'net')
>>> net.lib
<msl.loadlib.load_library.DotNetContainer object at 0x7fe5e8372dd8>
>>> net.lib.Microsoft.Dynamics.BusinessConnectorNet.Axapta
<class 'Microsoft.Dynamics.BusinessConnectorNet.Axapta'>
```
I used the _standard_ Dependency Walker and [DependencyWalker for .NET](https://github.com/isindicic/DependencyWalker.Net) to create a file that contains the path to the dependency that these two Walker's found for `Microsoft.Dynamics.BusinessConnectorNet.dll`.

I created a single directory in Windows which contained 301 dependencies and then I copied this directory to a `/home/joe/issue6/` directory in Linux which only contained the `Microsoft.Dynamics.BusinessConnectorNet.dll` file. Therefore, in my `/home/joe/issue6/` directory there are 302 DLL's.

The summary of where on my computer the 301 dependencies came from can be found [here](https://github.com/MSLNZ/msl-loadlib/files/1712658/Microsoft.Dynamics.BusinessConnectorNet-dependencies.txt). My hope is that your computer is missing >= 1 of these dependencies and that you can find those missing DLL's on your Windows Server 2016 machine.

**Comment by denfromufa** [comment]
@ZhukovGreen  @jborbely I'm not 100% sure what is going on but i got another error with your repo:

```
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper>python test_main.py
.
Microsoft.Dynamics.BusinessConnectorNet.dll
Traceback (most recent call last):
  File "test_main.py", line 21, in <module>
    test_msl_connector()
  File "test_main.py", line 14, in test_msl_connector
    connectorLL = LoadLibrary(DLL_PATH, 'net')
  File "C:\Python\Python35_64b\lib\site-packages\msl_loadlib-0.3.2-py3.5.egg\msl\loadlib\load_library.py", line 66, in __init__
    if not os.path.splitext(_path)[1]:
  File "C:\Python\Python35_64b\lib\ntpath.py", line 224, in splitext
    return genericpath._splitext(p, '\\', '/', '.')
  File "C:\Python\Python35_64b\lib\genericpath.py", line 118, in _splitext
    sepIndex = p.rfind(sep)
AttributeError: 'WindowsPath' object has no attribute 'rfind'
```

**Comment by jborbely** [comment]
@denfromufa Based on your comment, it looks like the command you executed was `python test_main.py` and you got that `AttributeError`. I find that odd since running `python test_main.py` would only import 2 modules, define a `DLL_PATH` constant and define a `test_msl_connector()` function, but that function does not actually get called and so running `python test_main.py` does essentially nothing. The only possible error could have been an `ImportError`.

Sorry, I'm not replying to your comment with any useful or relevant information, I was just curious why _something_ happened.

**Comment by denfromufa** [comment]
I added main block to call that function:

if __name__=="__main__":
test_msl_connector()

**Comment by jborbely** [comment]
@denfromufa Thanks for the clarification.

If you have time, could you insert `DLL_PATH = DLL_PATH.as_posix()` in `test_main.py` and try again?

You will probably receive a 
> OSError: The library appears to be from a .NET Framework version < 4.0...

warning, but just run the script again as this relates to the `useLegacyV2RuntimeActivationPolicy="true"` XML attribute that needs to be created.

The `AttributeError` relates to the fact that the `pathlib` module was used to define the file path and you are using Python 3.5 to run the script. Apparently, in Python 3.6 the `os.path.splitext()` function seems to be okay receiving a `<class 'pathlib.WindowsPath'>` object.

**Comment by ZhukovGreen** [comment]
@denfrimufa maybe just add str(DLL_PATH)

**Comment by ZhukovGreen** [comment]
@denfromufa try use str(DLL_PATH). Pathlib was less evolved at python 3.5

**Comment by denfromufa** [comment]
@ZhukovGreen ok, i installed python 3.6 64-bit to test your pathlib usage :) never tried pathlib before, i feel its syntax unreadable for newcomers, especially from other languages. Anyway see my full CMD session below. In general there is no issues with venv, but be careful to activate it from its root directory, not from parent python directory.

<details>
  <summary>Click to expand python code</summary>
 
```python
C:\Python\Python36-64>python C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\Microsoft.Dynamics.BusinessConnectorNet.dll
<class 'Microsoft.Dynamics.BusinessConnectorNet.Axapta'>

C:\Python\Python36-64>python -m venv C:\Python\Python36-64\envs\testing

C:\Python\Python36-64>.\envs\testing\Scripts\activate.bat
(testing) C:\Python\Python36-64>where python
C:\Python\Python36-64\python.exe
C:\Python\Python36-64\envs\testing\Scripts\python.exe
C:\Python\Python35_64b\Scripts\python
C:\Python\Python35_64b\python.exe

(testing) C:\Python\Python36-64>where pip
C:\Python\Python36-64\envs\testing\Scripts\pip.exe
C:\Python\Python35_64b\Scripts\pip.exe

(testing) C:\Python\Python36-64>deactivate
C:\Python\Python36-64>cd envs\testing

C:\Python\Python36-64\envs\testing>.\Scripts\activate.bat
(testing) C:\Python\Python36-64\envs\testing>where python
C:\Python\Python36-64\envs\testing\Scripts\python.exe
C:\Python\Python35_64b\Scripts\python
C:\Python\Python35_64b\python.exe

(testing) C:\Python\Python36-64\envs\testing>where pip
C:\Python\Python36-64\envs\testing\Scripts\pip.exe
C:\Python\Python35_64b\Scripts\pip.exe

(testing) C:\Python\Python36-64\envs\testing>python C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py
Traceback (most recent call last):
  File "C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py", line 4, in <module>
    from msl.loadlib import LoadLibrary
ModuleNotFoundError: No module named 'msl'

(testing) C:\Python\Python36-64\envs\testing>pip install msl-loadlib
Collecting msl-loadlib
  Using cached msl_loadlib-0.3.2-py2.py3-none-win_amd64.whl
Collecting pythonnet>=2.3 (from msl-loadlib)
  Using cached pythonnet-2.3.0-cp36-cp36m-win_amd64.whl
Installing collected packages: pythonnet, msl-loadlib
Successfully installed msl-loadlib-0.3.2 pythonnet-2.3.0

(testing) C:\Python\Python36-64\envs\testing>python C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\Microsoft.Dynamics.BusinessConnectorNet.dll
Traceback (most recent call last):
  File "C:\Python\Python36-64\envs\testing\lib\site-packages\msl\loadlib\load_library.py", line 100, in __init__
    self._assembly = clr.System.Reflection.Assembly.LoadFile(self._path)
System.IO.FileLoadException: Mixed mode assembly is built against version 'v2.0.50727' of the runtime and cannot be loaded in the 4.0 runtime without additional configuration information.
   at System.Reflection.RuntimeAssembly.nLoadFile(String path, Evidence evidence)
   at System.Reflection.Assembly.LoadFile(String path)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py", line 21, in <module>
    Axapta = test_msl_connector()
  File "C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py", line 14, in test_msl_connector
    connectorLL = LoadLibrary(DLL_PATH, 'net')
  File "C:\Python\Python36-64\envs\testing\lib\site-packages\msl\loadlib\load_library.py", line 115, in __init__
    raise IOError(msg)
OSError: The library appears to be from a .NET Framework version < 4.0.
The useLegacyV2RuntimeActivationPolicy property was added to C:\Python\Python36-64\envs\testing\Scripts\python.exe.config
to fix the "System.IO.FileLoadException: Mixed mode assembly..." error.
Rerun the script, or shutdown and restart the interactive console, to see
if Python can now load the .NET library.


(testing) C:\Python\Python36-64\envs\testing>python C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\test_main.py
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper
C:\Users\denis\Downloads\MicrosoftDynamicsConnector_python_wrapper\Microsoft.Dynamics.BusinessConnectorNet.dll
<class 'Microsoft.Dynamics.BusinessConnectorNet.Axapta'>

(testing) C:\Python\Python36-64\envs\testing>
```
</details>

**Comment by denfromufa** [comment]
modified Python code:

```python
"""Implements wrapper for axapta bussiness connector."""

import pathlib
from msl.loadlib import LoadLibrary
import sys
print(pathlib.Path(__file__).parent)
sys.path.append(pathlib.Path(__file__).parent)
DLL_PATH = pathlib.Path(
    __file__).parent / 'Microsoft.Dynamics.BusinessConnectorNet.dll'
print(DLL_PATH)

def test_msl_connector():
    """Get Axapta object via msl-loadlib package."""
    connectorLL = LoadLibrary(DLL_PATH, 'net')
    Microsoft = getattr(connectorLL.lib,
                        'Microsoft.Dynamics.BusinessConnectorNet')
    Axapta = Microsoft.Dynamics.BusinessConnectorNet.Axapta
    return Axapta

if __name__=="__main__":
	Axapta = test_msl_connector()
	print(Axapta)
```

**Comment by jborbely** [comment]
@denfromufa Thanks for taking the time to run the test. Hopefully @ZhukovGreen can resolve his issue by finding the missing [dependency](https://github.com/MSLNZ/msl-loadlib/issues/6#issuecomment-364593856)

**Comment by jborbely** [comment]
@ZhukovGreen Have you been able to solve your dependency issue?

**Comment by ZhukovGreen** [comment]
@jborbely I'll let you know in a few days. But there is no chance to fail since you made it work on Linux:)
@denfromufa Hi Denis, have you managed to run the tests?

**Comment by ZhukovGreen** [comment]
@jborbely I confirm that the problem resolved, and I guess it has nothing common with your package. Thus closing the issue?

**Comment by jborbely** [comment]
@ZhukovGreen Yes, I will close this issue.

I have released a new version today (v0.4.0) that resolves all of the issues that were discovered while investigating the cause of your issue with dependencies, plus the release includes some new features. 

_**NOTE:**_ The way that you were accessing the Axapta class in v0.3.2
```python
getattr(connectorLL.lib, 'Microsoft.Dynamics.BusinessConnectorNet')
```
will no longer work in this release because you will get an `AttributeError` if you do not update your code.

With this release you would do the following to access the Axapta class

```python
>>> from msl.loadlib import LoadLibrary
>>> net = LoadLibrary('Microsoft.Dynamics.BusinessConnectorNet.dll', 'net')
>>> Axapta = net.lib.Microsoft.Dynamics.BusinessConnectorNet.Axapta
```
If you do have any issues with the latest release please create a new issue.
